### PR TITLE
add: add proxies option to pass traffic through burp suite

### PR DESCRIPTION
**Pull request by arthur4ires**
Hi, hope you are well :)

Thanks a lot for the tool, it contributes a modification to watch traffic using BurpSuite.

The example command is as follows:

```
python3 -m clairvoyance -vv -o output -w google-10000-english.txt -p http://127.0.0.1:8080 -k https://site/graphql
```

### CODE REVIEW

File: clairvoyance/__main__.py [hunk 1]
```diff
@@ -55,6 +55,14 @@ def parse_args():
         type=argparse.FileType("r"),
         help="This wordlist will be used for all brute force effots (fields, arguments and so on)",
     )
+    parser.add_argument(
+        "-p",
+        "--proxy",
+        dest="proxies",
+        default=False,
+        required=False,
+        help="Use a proxy such as BurpSuite to debug requests.",
```
Comment (nikitastupin) [inline clairvoyance/__main__.py:64]:
```suggestion
        help="Use the specified HTTP proxy (e.g. BurpSuite to debug requests). This parameter is passed to https://requests.readthedocs.io/en/latest/user/advanced/#proxies.",
```

File: clairvoyance/__main__.py [hunk 2]
```diff
@@ -55,6 +55,14 @@ def parse_args():
         type=argparse.FileType("r"),
         help="This wordlist will be used for all brute force effots (fields, arguments and so on)",
     )
+    parser.add_argument(
+        "-p",
+        "--proxy",
+        dest="proxies",
+        default=False,
+        required=False,
```
Comment (nikitastupin) [inline clairvoyance/__main__.py:63]:
```suggestion
        metavar="[protocol://]host[:port]",
```

File: clairvoyance/__main__.py [hunk 3]
```diff
@@ -55,6 +55,14 @@ def parse_args():
         type=argparse.FileType("r"),
         help="This wordlist will be used for all brute force effots (fields, arguments and so on)",
     )
+    parser.add_argument(
+        "-p",
+        "--proxy",
+        dest="proxies",
+        default=False,
```
Comment (nikitastupin) [inline clairvoyance/__main__.py:62]:
```suggestion
        default=None,
```

**Comment by arthur4ires** [comment]
For me it's ok, thank you for the reply :)

File: clairvoyance/__main__.py [hunk 4]
```diff
@@ -77,6 +85,12 @@ def parse_args():
     config = graphql.Config()
     config.url = args.url
     config.verify = not args.insecure
+    
+    if args.proxies != False:
+        config.proxies = args.proxies
+    else:
+        config.proxies = False
+
```
Comment (nikitastupin) [inline clairvoyance/__main__.py:93]: Can be simplified (see the comment above)

File: clairvoyance/graphql.py [hunk 5]
```diff
@@ -34,7 +36,10 @@ def post(url, data=None, json=None, **kwargs):
 
     session.mount("http://", adapter)
     session.mount("https://", adapter)
-
+    
+    if proxies != None:
+        session.proxies = {"http": proxies, "https": proxies}
+
```
Comment (nikitastupin) [inline clairvoyance/graphql.py:42]: Can be simplified also

**Review by nikitastupin — CHANGES_REQUESTED** [review]
Hey @arthur4ires!

Thank you for the contribution 😃 

I've left a few comments I'd like to address or discuss before merging.

**Comment by arthur4ires** [comment]
Yes, I will make the changes :)

**Comment by arthur4ires** [comment]
Hello @nikitastupin , I accepted your modifications, thanks for the suggestions in the code!

File: clairvoyance/graphql.py [hunk 6]
```diff
@@ -12,7 +13,8 @@
 from typing import Set
 
 
-def post(url, data=None, json=None, **kwargs):
+def post(url, data=None, json=None, proxies=None, **kwargs):
```
Comment (nikitastupin) [inline clairvoyance/graphql.py:16]: I suppose the `proxies` can be passed as part of `**kwargs` so we don't need the explicit `proxies` parameter.

File: clairvoyance/__main__.py (hunk 4 above)
Comment (nikitastupin) [reply clairvoyance/__main__.py:93]:
```suggestion
    config.proxies = {"http": args.proxies, "https": args.proxies} if args.proxies else None
        
```

File: clairvoyance/graphql.py (hunk 6 above)
Comment (nikitastupin) [reply clairvoyance/graphql.py:16]: The same way as `verify` is passed

File: clairvoyance/oracle.py [hunk 7]
```diff
@@ -70,13 +70,26 @@ def probe_valid_fields(
...
+            
+        else:
+            
+            response = graphql.post(
+                config.url,
+                headers=config.headers,
+                json={"query": document},
+                verify=config.verify,
+            )
+
```
[the remark this answers is not in the document]
Comment (nikitastupin) [reply clairvoyance/oracle.py:92]: There are also other `.post(...)` calls, we should address them aswell

**Review by nikitastupin — CHANGES_REQUESTED** [review]
Thanks! I think we can further simplify the code and also add proxies to all requests. Would you be willing to make the changes I've requested in comments?

**Comment by zeroc00I** [comment]
Sup folks
Some updates here?

**Comment by nikitastupin** [comment]
Hey @zeroc00I,

I believe you can use the `HTTP_PROXY` & `HTTPS_PROXY` environment variables to proxy clairvoyance https://requests.readthedocs.io/en/latest/user/advanced/#proxies

**Comment by QuentinN42** [comment]
Hey @arthur4ires @zeroc00I at Escape, we are working to secure the graphql api. We forked this repo a week ago and I implemented this feature there.

https://github.com/Escape-Technologies/ClairvoyanceNext

**Comment by arthur4ires** [comment]
Hi there, thanks a lot for the update :)

**Pull request closed without merging**